core/trino-main/src/main/java/io/trino/execution/SetColumnTypeTask.java

```diff
@@ -94,16 +101,53 @@ else if (metadata.getView(session, qualifiedObjectName).isPresent()) {
 
         TableHandle tableHandle = redirectionAwareTableHandle.getTableHandle().get();
         Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
-        ColumnHandle column = columnHandles.get(statement.getColumnName().getValue().toLowerCase(ENGLISH));
+        String columnName = statement.getColumnName().getParts().get(0).toLowerCase(ENGLISH);
+        ColumnHandle column = columnHandles.get(columnName);
         if (column == null) {
             throw semanticException(COLUMN_NOT_FOUND, statement, "Column '%s' does not exist", statement.getColumnName());
         }
 
-        metadata.setColumnType(session, tableHandle, column, getColumnType(statement));
+        Type type = getColumnType(statement);
+        if (statement.getColumnName().getParts().size() == 1) {
+            metadata.setColumnType(session, tableHandle, column, type);
+        }
+        else {
+            ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle, column);
+            List<String> fieldPath = statement.getColumnName().getParts();
+
+            Type currentType = columnMetadata.getType();
+            for (int i = 1; i < fieldPath.size(); i++) {
+                String fieldName = fieldPath.get(i);
+                List<RowType.Field> candidates = getCandidates(currentType, fieldName);
+
+                if (candidates.isEmpty()) {
+                    throw semanticException(COLUMN_NOT_FOUND, statement, "Field '%s' does not exist within %s", fieldName, currentType);
+                }
+                if (candidates.size() > 1) {
+                    throw semanticException(AMBIGUOUS_NAME, statement, "Field path %s within %s is ambiguous", fieldPath, columnMetadata.getType());
+                }
+                currentType = getOnlyElement(candidates).getType();
+            }
+
+            metadata.setFieldType(session, tableHandle, fieldPath, type);
+        }
 
         return immediateVoidFuture();
     }
 
+    private static List<RowType.Field> getCandidates(Type type, String fieldName)
+    {
+        if (!(type instanceof RowType rowType)) {
+            throw new TrinoException(NOT_SUPPORTED, "Unsupported type: " + type);
+        }
+        List<RowType.Field> candidates = rowType.getFields().stream()
+                // case-insensitive match
```

Comment:
Can this be case sensitive? (eg not to get us one step further from #17)

Reply:
This is the same logic as row fields resolution in expressions as you may know.  
@kasiafi requested a case-insensitive match in adding fields PR https://github.com/trinodb/trino/pull/16321#discussion_r1169956313. 

Currently, adding & renaming fields are case-insensitive and dropping field (this was the 1st PR about managing nested fields) is case-sensitive. We should change dropping fields to case-insensitive in the near future.